controllers/application/controller.go

```diff
@@ -109,9 +107,59 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req reconcile.Req
 		return reconcile.Result{}, err
 	}
 
+	tmpApplication := application.DeepCopy()
+	application.FillDefaultsSpec()
+	if !ctrlutil.ContainsFinalizer(application, applicationFinalizer) {
+		ctrlutil.AddFinalizer(application, applicationFinalizer)
+	}
+
+	if len(application.Labels) == 0 {
+		application.Labels = application.Spec.Labels
+	} else {
+		aggregateLabels := application.Labels
+		maps.Copy(aggregateLabels, application.Spec.Labels)
+		application.Labels = aggregateLabels
+	}
+
+	application.FillDefaultsStatus()
+
+	specDiff, err := util.GetObjectDiff(tmpApplication.Spec, application.Spec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	statusDiff, err := util.GetObjectDiff(tmpApplication.Status, application.Status)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	fullDiff, err := util.GetObjectDiff(tmpApplication, application)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// If we update the Application initially on applied defaults before starting reconciling resources we allow all
+	// updates to be visible even though the controllerDuties may take some time.
+	if len(specDiff) > 0 {
+		err := r.GetClient().Update(ctx, application)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if len(statusDiff) > 0 {
+		err := r.GetClient().Status().Update(ctx, application)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if len(fullDiff) > 0 && len(statusDiff) == 0 {
+		err := r.GetClient().Update(ctx, application)
+		return reconcile.Result{Requeue: true}, err
+	}
```

Comment:
this should not be necessary, but for some reason L#143-L#146 does not apply the finalizer...